pkg/controller/starter.go

```diff
@@ -0,0 +1,81 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	appsv1 "k8s.io/client-go/kubernetes/typed/apps/v1"
+	"k8s.io/klog"
+
+	"github.com/openshift/library-go/pkg/controller/controllercmd"
+	"github.com/openshift/service-ca-operator/pkg/controller/api"
+	apistart "github.com/openshift/service-ca-operator/pkg/controller/apiservicecabundle/starter"
+	cmstart "github.com/openshift/service-ca-operator/pkg/controller/configmapcainjector/starter"
+	certstart "github.com/openshift/service-ca-operator/pkg/controller/servingcert/starter"
+	"github.com/openshift/service-ca-operator/pkg/operator/operatorclient"
+)
+
+func StartServiceCAControllers(ctx context.Context, controllerContext *controllercmd.ControllerContext) error {
+	// Maybe scale down the managing deployment if a downgrade is indicated by the
+	// existence of legacy independent deployments.
+	//
+	// It should be safe to check infrequently. The behavior of legacy deployments and
+	// the current deployment is intended to be compatible, so having them run
+	// simultaneously for short periods of time is unlikely to result in hot loops.
+	kubeClient, err := kubernetes.NewForConfig(controllerContext.ProtoKubeConfig)
+	if err != nil {
+		return fmt.Errorf("error creating kube client: %v", err)
+	}
+	deployClient := kubeClient.AppsV1().Deployments(operatorclient.TargetNamespace)
+	go wait.Until(func() { maybeScaleDown(deployClient) }, 30*time.Minute, ctx.Done())
+
+	err = apistart.StartAPIServiceCABundleInjector(ctx, controllerContext)
+	if err != nil {
+		return err
+	}
+	err = cmstart.StartConfigMapCABundleInjector(ctx, controllerContext)
+	if err != nil {
+		return err
+	}
+	err = certstart.StartServiceServingCertSigner(ctx, controllerContext)
+	if err != nil {
+		return err
+	}
+
+	<-ctx.Done()
+
+	return fmt.Errorf("stopped")
+}
+
+// maybeScaleDown scales down the deployment managing this pod if a downgrade has
+// occurred to a legacy version of the operator that uses a deployment per
+// controller.
+func maybeScaleDown(deployClient appsv1.DeploymentInterface) {
```

Comment:
nit: can you add a note about when all of this is going to be safe to be removed?

Reply:
Done.

Reply:
I've added a note at the call point indicating when removal is an option.

Reply:
I meant an exact version - it should be safe for it to be removed in 4.5 since downgrade would still have to pass 4.4, no?

Reply:
Would it possible to upgrade from 4.2 to 4.5?  Would it be possible then to downgrade from 4.5 to 4.2? 

Also, can you point me to documentation on how upgrades work? I don't think I understand what upgrade strategies are supported for openshift.